core/arch/arm/plat-imx/imx_caam.c

```diff
@@ -50,6 +52,44 @@ static void caam_enable_clocks(bool enable)
 
 }
 
+static void imx_caam_reset_jr(struct imx_caam_ctrl *caam)
+{
+	uint32_t reg_val;
+	uint32_t timeout = 1000;
+
+	io_write32((vaddr_t)&caam->jrcfg[0].jrcr, 1);
+	do {
+		isb();
+		dsb();
+		reg_val = io_read32((vaddr_t)&caam->jrcfg[0].jrintr);
+		reg_val &= 0xc;
+	} while ((reg_val == 0x4) && --timeout);
+
+	io_write32((vaddr_t)&caam->jrcfg[0].jrcr, 1);
+	do {
+		isb();
+		dsb();
+		reg_val = io_read32((vaddr_t)&caam->jrcfg[0].jrintr);
+		reg_val &= 0xc;
+	} while ((reg_val & 0x1) && --timeout);
+}
+
+static struct imx_caam_jr *caam_init_jr(struct imx_caam_ctrl *caam)
+{
+	struct imx_caam_jr *jr = calloc(1, sizeof(struct imx_caam_jr));
+
+	imx_caam_reset_jr(caam);
+	jr->njobs = 4;
+	jr->inring = calloc(jr->njobs, sizeof(struct imx_caam_inring_entry));
+	jr->outring = calloc(jr->njobs,  sizeof(struct imx_caam_outring_entry));
+	io_write32((vaddr_t)&caam->jrstartr, 1);
+	io_write32((vaddr_t)&caam->jrcfg[0].irbar_ls, virt_to_phys(jr->inring));
+	io_write32((vaddr_t)&caam->jrcfg[0].irsr, jr->njobs);
+	io_write32((vaddr_t)&caam->jrcfg[0].orbar_ls,
+		   virt_to_phys(jr->outring));
+	io_write32((vaddr_t)&caam->jrcfg[0].orsr, jr->njobs);
+	return jr;
+}
```

Comment:
If I'm reading the code right, you don't need to modify job-ring ownership from non-secure to secure world to do your HUK stuff here, is that correct ?

Reply:
Follow on question.
What is the effect of this code on runtime Linux ?

Reply:
No, the jobring ownership modification is not necessary since the `PRIBLOB` bits are used to prevent another CAAM user from obtaining the MKVB.

There should be no effect on runtime Linux, since the CAAM is used once during OP-TEE startup and not used afterwards. However this is not the case with the current PR, since `huk_subkey_derive` expects `tee_otp_get_hw_unique_key` to be callable multiple times.
The solution to this is to save the MKVB and only retrieve it once, maybe during an earlier `init_call` which runs before the service init. This needs to be implemented before this PR is ready. (I just dug into this after looking at the system PTA which also uses `huk_subkey_derive`)

Reply:
*sigh* Reading the code further: the `HUK` evidently is not supposed to be retained in memory, since its zeroed explicitly. I guess the attack scenario is someone being able to read the secure memory and thus being able to obtain all encryption keys, but not being able to compomise the `HUK`.

Reply:
Do you reckon that's a realistic attack ?
From my POV so long as your HUK code doesn't break runtime Linux with job-ring permission conflict it looks fine. So if that means reading the HUK and storing in OPTEE memory my own view is that is not really a problem.

Reply:
There's a point in not keeping HUK in memory longer than necessary. Because as long as HUK isn't revealed the attacker is limited to tricking OP-TEE into doing unintended things with HUK instead of letting the attacker do whatever he likes with it.

Reply:
I can see the additional security benefit for other platforms, unfortunately this is not possible for the i.MX6 platform AFAICS. We need to increment the PRIBLOB field and can not generate the same MKVB again afterwards.